src/estimagic/inference/bootstrap_samples.py

```diff
@@ -19,17 +26,45 @@ def get_bootstrap_indices(data, rng, cluster_by=None, n_draws=1000):
 
     """
     n_obs = len(data)
-    if cluster_by is None:
-        bootstrap_indices = list(rng.integers(0, n_obs, size=(n_draws, n_obs)))
+
+    if weights is None:
+
+        if cluster_by is None:
+            bootstrap_indices = list(rng.integers(0, n_obs, size=(n_draws, n_obs)))
+        else:
+            clusters = data[cluster_by].unique()
+            drawn_clusters = rng.choice(
+                clusters, size=(n_draws, len(clusters)), replace=True
+            )
+
+            bootstrap_indices = _convert_cluster_ids_to_indices(
+                data[cluster_by], drawn_clusters
+            )
+
     else:
-        clusters = data[cluster_by].unique()
-        drawn_clusters = rng.choice(
-            clusters, size=(n_draws, len(clusters)), replace=True
-        )
 
-        bootstrap_indices = _convert_cluster_ids_to_indices(
-            data[cluster_by], drawn_clusters
-        )
+        if cluster_by is None:
+            bootstrap_indices = list(
+                rng.choice(
+                    n_obs,
+                    size=(n_draws, n_obs),
+                    replace=True,
+                    p=data[weights] / data[weights].sum(),
+                )
+            )
+        else:
+            clusters = data.groupby(cluster_by)[weights].sum().reset_index()
+
+            drawn_clusters = rng.choice(
+                clusters[cluster_by],
+                size=(n_draws, len(clusters)),
+                replace=True,
+                p=clusters[weights] / clusters[weights].sum(),
+            )
+
+            bootstrap_indices = _convert_cluster_ids_to_indices(
+                data[cluster_by], drawn_clusters
+            )
```

Comment:
This has to be updated. Especially the line `clusters[clusters_by]` is very confusing. I would think that `clusters` is equal to something like `data[cluster_by]`.

Similarly as above, it would be nice to pre-compute the probabilities.